octodns/provider/googlecloud.py

```diff
@@ -0,0 +1,348 @@
+        else:
+            gcloud_zone = self.gcloud_zones.get(desired.name)
+
+        gcloud_changes = gcloud_zone.changes()
+
+        for change in changes:
+            class_name = change.__class__.__name__
+            _rrset_func = getattr(
+                self, '_rrset_for_{}'.format(change.record._type))
+
+            if class_name == 'Create':
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Delete':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Update':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.existing))
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.new))
+            else:
+                raise RuntimeError('Change type "{}" for change "{!s}" '
+                                   'is none of "Create", "Delete" or "Update'
+                                   .format(class_name, change))
+
+        gcloud_changes.create()
+
+        for i in range(120):
+            gcloud_changes.reload()
+            self.log.debug("Waiting for changes to complete")
+            # https://cloud.google.com/dns/api/v1/changes#resource
+            # status can be one of either "pending" or "done"
+            if gcloud_changes.status != 'pending':
+                break
+            self.log.debug("Waiting for changes to complete")
+            time.sleep(self.CHANGE_LOOP_WAIT)
+
+        if gcloud_changes.status != 'done':
+            raise RuntimeError("Timeout reached after {} seconds".format(
+                i * self.CHANGE_LOOP_WAIT))
+
+    def _create_gcloud_zone(self, dns_name):
+        """Creates a google cloud ManagedZone with dns_name, and zone named
+            derived from it. calls .create() method and returns it.
+
+            :param dns_name: fqdn of zone to create
+            :type  dns_name: str
+
+            :type return: new google.cloud.dns.ManagedZone
+        """
+        # Zone name must begin with a letter, end with a letter or digit,
+        # and only contain lowercase letters, digits or dashes
+        zone_name = re.sub("[^a-z0-9-]", "",
+                           dns_name[:-1].replace('.', "-"))
+
+        # Check if there is another zone in google cloud which has the same
+        # name as the new one
+        while zone_name in [z.name for z in self.gcloud_zones.values()]:
+            # If there is a zone in google cloud alredy, then try suffixing the
+            # name with a -i where i is a number which keeps increasing until
+            # a free name has been reached.
+            m = re.match("^(.+)-([0-9]+$)", zone_name)
+            if m:
+                i = int(m.group(2)) + 1
+                zone_name = "{}-{!s}".format(m.group(1), i)
+            else:
+                zone_name += "-2"
```

Comment:
What case is this having to handle with the appends? Wouldn't it be sufficient to make the name be dns_name.replace('.', '-')? It'd seem really confusing to have multiple things with the same `zone_name`, but different `dns_names`. Not sure how things would get into that situation.

Reply:
The rationale is that since `dns_names` may have both hyphens and dots in the name whereas `zone_name` can have only hyphens, there are situations in which dns_name.replace('.', '-') of two differet dns_names results the same zone_name (for example say forest-treck.example.com and forest.treck.example.com).

Reply:
OK. Weird that google cloud has that sort of display name and it doesn't just use the uniq dns_name. I assume leaking some sort of internal implementation detail or doing it for for consistency with something somewhere else. :slightly_smiling_face: 

Anyway, given the full context I think the best bet here is to just use a uuid on all names which it's safe to assume will be unique without having to repeatedly check, and it can just always do it. There's a similar bit in the DynProvider for [its `DSFRuleset.label`](https://github.com/github/octodns/blob/e0f357d9f9c041238b069b5fc4c1f4c663bf16a8/octodns/provider/dyn.py#L545). 

That'd get rid of all the looping and regexes and just look like:

```python
zone_name = '{}-{}'.format(dns_name[:-1].replace('.', '-'), uuid4().hex)
```

Reply:
You may in google cloud create several dns zones with the same DNS name (but different zone names). They appear to put different NS records in the zone. I have yet to find out what is the benefit of this approach.